src/bedrock/monitor/monitoring.cpp

```diff
@@ -946,246 +257,45 @@ int main(int argc, char* argv[]) {
-            }
-          }
-
-          change_replication_factor(requests, global_hash_ring_map, local_hash_ring_map, routing_address, placement, pushers, mt, response_puller, logger, rid);
-
-          ServerThread node = ServerThread(ss.min_occupancy_memory_ip, 0);
-          auto connection_addr = node.get_self_depart_connect_addr();
-
-          departing_node_map[ss.min_occupancy_memory_ip] = tier_data_map[1].thread_number_;
-          auto ack_addr = mt.get_depart_done_connect_addr();
-
-          logger->info("Removing memory node {}.", node.get_ip());
-          zmq_util::send_string(ack_addr, &pushers[connection_addr]);
-          removing_memory_node = true;
-        }
-      }
-
-      requests.clear();
-
-      // finally, consider reducing the replication factor of some keys that are not so hot anymore
-      for (auto it = key_access_summary.begin(); it != key_access_summary.end(); it++) {
-        string key = it->first;
-        unsigned total_access = it->second;
-
-        if (!is_metadata(key) && total_access <= ss.key_access_mean && placement[key].global_replication_map_[1] > MINIMUM_REPLICA_NUMBER) {
-          logger->info("Key {} accessed {} times (threshold is {}).", key, total_access, ss.key_access_mean);
-          requests[key] = create_new_replication_vector(1, MINIMUM_REPLICA_NUMBER - 1, 1, 1);
-          logger->info("Dereplication for key {}. M: {}->{}. E: {}->{}", key, placement[key].global_replication_map_[1], requests[key].global_replication_map_[1], placement[key].global_replication_map_[2], requests[key].global_replication_map_[2]);
-        }
-      }
-
-      change_replication_factor(requests, global_hash_ring_map, local_hash_ring_map, routing_address, placement, pushers, mt, response_puller, logger, rid);
-      requests.clear();
-
-      logger->info("Adding {} memory nodes is in progress.", adding_memory_node);
-      logger->info("Adding {} EBS nodes in progress.", adding_ebs_node);
+      policy_algorithm(logger,
+                       global_hash_ring_map,
+                       local_hash_ring_map,
+                       grace_start,
+                       ss,
+                       adding_memory_node,
+                       adding_ebs_node,
+                       removing_memory_node,
+                       removing_ebs_node,
+                       management_address,
+                       placement,
+                       key_access_summary,
+                       mt,
+                       tier_data_map,
+                       departing_node_map,
+                       pushers,
+                       response_puller,
+                       routing_address,
+                       rid,
+                       rep_factor_map,
+                       MINIMUM_MEMORY_NODE,
```

Comment:
These values shouldn't be passed in to the method. This is why we went through the whole song-and-dance of defining them as `extern consts` in `common.hpp`. You should be able to define and set them in this file and use them globally.

Reply:
fixed in the latest commit